dask_expr/_reductions.py

```diff
@@ -65,58 +85,140 @@ def _layer(self):
+
+
+class TreeReduce(Expr):
+    """Tree-reduction component of `ApplyConcatApply`
+
+    This class is used within `ApplyConcatApply._lower`.
+
+    See Also
+    --------
+    ApplyConcatApply
+    """
+
+    _parameters = [
+        "frame",
+        "kind",
+        "_meta",
+        "combine",
+        "aggregate",
+        "combine_kwargs",
+        "aggregate_kwargs",
+    ]
+
+    def __dask_postcompute__(self):
+        return toolz.first, ()
+
+    def _layer(self):
         # apply combine to batches of intermediate results
+        j = 1
+        d = {}
+        keys = self.frame.__dask_keys__()
+        split_every = getattr(self, "split_every", 0)
         while len(keys) > 1:
             new_keys = []
             for i, batch in enumerate(
                 toolz.partition_all(split_every or len(keys), keys)
             ):
                 batch = list(batch)
-                if combine_kwargs:
-                    d[self._name, j, i] = (apply, combine, [batch], combine_kwargs)
+                if self.combine_kwargs:
+                    d[self._name, j, i] = (
+                        apply,
+                        self.combine,
+                        [batch],
+                        self.combine_kwargs,
+                    )
                 else:
-                    d[self._name, j, i] = (combine, batch)
+                    d[self._name, j, i] = (self.combine, batch)
                 new_keys.append((self._name, j, i))
             j += 1
             keys = new_keys
 
         # apply aggregate to the final result
-        d[self._name, 0] = (apply, aggregate, [keys], aggregate_kwargs)
+        d[self._name, 0] = (apply, self.aggregate, [keys], self.aggregate_kwargs)
 
         return d
 
     @property
     def _meta(self):
-        meta = meta_nonempty(self.frame._meta)
-        meta = self.chunk(meta, **self.chunk_kwargs)
-        aggregate = self.aggregate or (lambda x: x)
-        if self.combine:
-            combine = self.combine
-            combine_kwargs = self.combine_kwargs
-        else:
-            combine = aggregate
-            combine_kwargs = self.aggregate_kwargs
-
-        meta = combine([meta], **combine_kwargs)
-        meta = aggregate([meta], **self.aggregate_kwargs)
-        return make_meta(meta)
+        return self.operand("_meta")
 
     def _divisions(self):
         return (None, None)
 
+    def __str__(self):
+        chunked = str(self.frame)
+        split_every = getattr(self, "split_every", 0)
+        return f"{type(self).__name__}({chunked}, kind={funcname(self.kind)}, split_every={split_every})"
+
+    def _tree_repr_lines(self, indent=0, recursive=True):
+        header = f"{funcname(self.kind)}({funcname(type(self))}): "
+        lines = []
+        if recursive:
+            for dep in self.dependencies():
+                lines.extend(dep._tree_repr_lines(2))
+
+        split_every = getattr(self, "split_every", 0)
+        header += f" split_every={split_every}"
```

Comment:
Yeah, my thought was that the most important parameter for the `TreeReduce` component is `split_every`, because everything else will just match the `Chunk` dependency. I'm very open to other solutions here.